---
title: "Bach single-cell mammary RNA-seq"
output: html_notebook
---

```{r}
rm(list=ls())

### PACKAGES
library(gplots)
library(ggplot2)
library(stats)
library(matrixStats)
library(grDevices)
library(RColorBrewer)
library(Matrix)
library(matrixStats)

### DIRECTORIES
root_dir = "/Users/adelexu/Google Drive/PhD/S27L_manuscript/source_data/analysis_plotting_code"
project_dir = file.path(root_dir, "bach_single_cell_mammary")
input_dir = file.path(project_dir, "data")
output_dir = file.path(project_dir, "results")

### INPUT FILES

gene_annotation_file = file.path(input_dir, "gene_annotation_GRCm38.88.csv")

# These files were downloaded from GSE106273
l1_umi_file = file.path(input_dir, "GSM2834502_L_1_matrix.mtx")
l2_umi_file = file.path(input_dir, "GSM2834503_L_2_matrix.mtx")

l1_barcode_file = file.path(input_dir, "GSM2834502_L_1_barcodes.tsv")
l2_barcode_file = file.path(input_dir, "GSM2834503_L_2_barcodes.tsv")

l1_genes_file = file.path(input_dir, "GSM2834502_L_1_genes.tsv")
l2_genes_file = file.path(input_dir, "GSM2834503_L_2_genes.tsv")

rp_annotation_file = file.path(input_dir, "rp_annotation.tsv")

```

```{r}
### READ INPUT

gene_annotation = read.csv(gene_annotation_file, header=TRUE)

# separate out mitochondrial genes
mito_genes = gene_annotation[gene_annotation$Chromosome.scaffold.name=="MT","Gene.stable.ID"] #37 genes, consistent w/ expectations -- but 2 are rRNAs and 22 are tRNAs, so only expect 13 mRNAs

l1_umi = as.matrix(readMM(l1_umi_file)) #124.8 Mb unzipped, took < 1 min to read. returns Large dgTMatrix, 152.7 Mb --> if coerced to regular matrix by as.matrix, 1.2 Gb
dim(l1_umi) # 27998  5906
l1_barcodes = paste("l1.", read.table(l1_barcode_file)$V1, sep="") #5906 entries
l1_genes = read.table(l1_genes_file) #27998 entries, gene_annotation IDs & gene names
colnames(l1_genes) = c("ensembl", "gene_name")

l2_umi = as.matrix(readMM(l2_umi_file))
dim(l2_umi) # 27998  3697
l2_barcodes = paste("l2.", read.table(l2_barcode_file)$V1, sep="") #3697 entries
l2_genes = read.table(l1_genes_file) #27998 entries, gene_annotation IDs & gene names
colnames(l2_genes) = c("ensembl", "gene_name")

rp_annotation = read.table(file=rp_annotation_file, header=TRUE, sep='\t')
```

```{r}
# format and combine inputs

rownames(l1_umi) = l1_genes$ensembl
colnames(l1_umi) = l1_barcodes
l1_umi[1:5, 1:5]

rownames(l2_umi) = l2_genes$ensembl
colnames(l2_umi) = l2_barcodes
l2_umi[1:5, 1:5]

l12_genes = l1_genes #l1_genes and l2_genes are equivalent
l12_umi = cbind(l1_umi, l2_umi)
dim(l12_umi) #27998 9603

l12_barcodes = c(l1_barcodes, l2_barcodes)
rm(l1_genes, l2_genes, l1_umi, l2_umi, l1_barcodes, l2_barcodes)
```

```{r}
# double-check QC criteria reported in Bach et al:
# > 500 total genes
# > 1000 UMIs
# < 5% mitochondrial 

l12_total_genes = colSums(l12_umi > 0)
hist(l12_total_genes) # peak around 1000, long tail up to ~6500 genes
range(l12_total_genes) # 604 6549

l12_total_umis = colSums(l12_umi)
hist(l12_total_umis) # very long R side tail
range(l12_total_umis) # 1724 52835 -- now 57585, why?

l12_umi_mito_genes = l12_umi[rownames(l12_umi) %in% mito_genes,]
dim(l12_umi_mito_genes) #13 x 9603
l12_pc_mito = colSums(l12_umi_mito_genes)/l12_total_umis # total mito UMIs / total UMIs
range(l12_pc_mito) # 0.0000000 0.08217405

# some cells are over 5% mitochondrial -- need to remove from l12_umi and barcodes
cells_to_remove = l12_pc_mito > 0.05
sum(cells_to_remove) # only 1 cell to remove
l12_umi = l12_umi[,!cells_to_remove]
dim(l12_umi) #27998 9602
l12_barcodes = l12_barcodes[!cells_to_remove]
sum(l12_barcodes==colnames(l12_umi))
```

```{r}
# clustering cell types based on genes from Bach et al. Figure 2

avdl_markers = c("Olah", "Glycam1", "Thrsp", "Fabp3", "Rspo1", "Atp6v1c2", "Csn2", "Wap")
avpl_markers = c("Aldh1a3", "Kit", "Cd14", "Lypd3")
myo_markers = c("Igfbp2", "Oxtr")
prc_markers = c("Notch3", "Zeb2")
all_markers = unique(c(avdl_markers, avpl_markers, myo_markers, prc_markers))
length(all_markers) #16

all_markers %in% l12_genes$gene_name # all true
all_markers_gene_annotation = factor(all_markers, levels=l12_genes$gene_name, labels=l12_genes$ensembl) # checked that the right gene_annotation IDs were paired up w/ the marker genes

# format trimmed UMI matrix
l12_umi_markers = l12_umi[rownames(l12_umi) %in% all_markers_gene_annotation,]
dim(l12_umi_markers) # 16 9602
rownames(l12_umi_markers) = as.vector(factor(rownames(l12_umi_markers), levels=l12_genes$ensembl, labels=l12_genes$gene_name))

# normalize, log transform, scale to 0-1 per gene
l12_umi_markers_heatmap = matrix(data=0, nrow=nrow(l12_umi_markers), ncol=ncol(l12_umi_markers))
rownames(l12_umi_markers_heatmap) = rownames(l12_umi_markers)
colnames(l12_umi_markers_heatmap) = colnames(l12_umi_markers)
for (i in 1:ncol(l12_umi_markers_heatmap)) {
  if (sum(l12_umi_markers[,i]) > 0) {
    l12_umi_markers_heatmap[,i] = l12_umi_markers[,i]/sum(l12_umi_markers[,i])*1000000
  }
}

l12_umi_markers_heatmap = log(l12_umi_markers_heatmap+1, 2)
for (j in 1:nrow(l12_umi_markers_heatmap)) {
  if (max(l12_umi_markers_heatmap[j,]) > 0) {
    l12_umi_markers_heatmap[j,] = l12_umi_markers_heatmap[j,]/max(l12_umi_markers_heatmap[j,])
  }
}

# heatmap.2(l12_umi_markers_heatmap, trace="none")

# heatmap.2(l12_umi_markers_heatmap[,!colMaxs(l12_umi_markers_heatmap)==0], distfun=function(x) as.dist(1-cor(t(x), method="spearman")), trace="none") # have to remove the cells that don't express any of the marker genes
# recall avd markers: "Olah", "Glycam1", "Thrsp", "Fabp3", "Rspo1", "Atp6v1c2", "Csn2", "Wap"

hist(l12_umi_markers_heatmap["Csn2",])
hist(l12_umi_markers_heatmap["Wap",])
hist(l12_umi_markers_heatmap["Fabp3",])

avp_score = colSums(l12_umi_markers_heatmap[c("Aldh1a3", "Kit", "Cd14"),])
hist(avp_score)

avd_score = colSums(l12_umi_markers_heatmap[c("Fabp3", "Thrsp", "Glycam1", "Olah"),])
hist(avd_score)

cell_annotation = data.frame(barcode=l12_barcodes)
cell_annotation$is_avp = avp_score > 1
cell_annotation$is_avd = avd_score > 2
cell_annotation$is_other = !(cell_annotation$is_avp | cell_annotation$is_avd) #9290
table(cell_annotation$is_avp, cell_annotation$is_avd)
table(cell_annotation$is_avp) #220 AVPs
table(cell_annotation$is_avd) #104 AVDs
# 6 overlap between AVPs and AVDs 
#        FALSE TRUE
#  FALSE  9290   92
#  TRUE    208   12
```

```{r}
# back to raw umi counts -- going to pool the counts for each cell type, then normalize per RP reads
rp_annotation = as.vector(rp_annotation[,"gene.id"]) # use all genes in rp_annotation
rp_annotation = rp_annotation[rp_annotation %in% rownames(l12_umi)] # Rps4y2 wasn't in the dataset

l12_avd_pooled = rowSums(l12_umi[rp_annotation,cell_annotation$is_avd])
l12_avd_pooled = l12_avd_pooled/sum(l12_avd_pooled)*100

rp_annotation_names = as.vector(factor(rp_annotation, levels=l12_genes$ensembl, labels=l12_genes$gene_name))

```

```{r}
#  normalize by total UMIs per cell
l12_umi_norm = matrix(data=0, nrow=nrow(l12_umi), ncol=ncol(l12_umi))
rownames(l12_umi_norm) = as.vector(factor(rownames(l12_umi), levels=l12_genes$ensembl, labels=l12_genes$gene_name))
colnames(l12_umi_norm) = colnames(l12_umi)
for (i in 1:ncol(l12_umi_norm)) {
  if (sum(l12_umi[,i]) > 0) {
    l12_umi_norm[,i] = l12_umi[,i]/sum(l12_umi[,i])*1000000
  }
}
```


```{r}
# calculate correlations for AVDs: Rps27l normalized umis vs other genes' umis

# calculate correlations based on Rps27l stoichiometry relative to RP genes, rather than normalized UMIs. OK to use l12_umi_norm for this, since the normalizing denominator cancels out anyway. For the other genes, still using normalized UMIs. Also not limiting to AVPs for now
l12_sum_all_rp_transcripts = colSums(l12_umi_norm[rownames(l12_umi_norm) %in% rp_annotation_names,])
l12_rps27l_stoichiometry = l12_umi_norm["Rps27l",]/l12_sum_all_rp_transcripts * 100 #normalized UMI for Rps27l divided by normalized UMI for all active genes

# among only avds
cor_rps27l_stoichiometry_avd = data.frame(gene_name = rownames(l12_umi_norm),
                                          spearman_rho = rep(NA, nrow(l12_umi_norm)),
                                          p_value = rep(NA, nrow(l12_umi_norm)))
for (i in 1:nrow(l12_umi_norm)) {
  if (i %% 1000 == 0) {
    print(i)
  }
  if (sum(l12_umi_norm[i,cell_annotation$is_avd]) > 0) {
    result = cor.test(l12_rps27l_stoichiometry[cell_annotation$is_avd], 
                      l12_umi_norm[i,cell_annotation$is_avd], 
                      alternative = "two.sided", 
                      method="spearman", 
                      exact = FALSE)
    cor_rps27l_stoichiometry_avd[i,] = c(rownames(l12_umi_norm)[i], result$estimate, result$p.value)
  }
}

```

```{r}
cor_rps27l_stoichiometry_avd$ensembl_id = l12_genes[,'ensembl']
write.csv(cor_rps27l_stoichiometry_avd[!is.na(cor_rps27l_stoichiometry_avd$spearman_rho),], file = file.path(output_dir, "S27L_stoich_correlations_to_norm_umi_spearman.csv"), quote=FALSE)
```

```{r}
# scatter plot for genes of interest

ygenes_oi = c("Rps27", "Csn2", "Wap", "Cdkn1a", "Mdm2")

for (ygene in ygenes_oi) {
  
  pdf(file=file.path(output_dir, paste("avd_rps27l_stoich_vs_", ygene, "_norm_umi_spearman", ".pdf", sep="")), width=4, height=4)
  df = data.frame(rps27l.stoich=l12_rps27l_stoichiometry[cell_annotation$is_avd], ygene.umi.norm=l12_umi_norm[ygene, cell_annotation$is_avd])
  par(pty='s')
  plot(x = df$rps27l.stoich,
       y = df$ygene.umi.norm,
       pch = 19,
       xlab="eS27L stoichiometry",
       ylab=ygene)
  # p0 = ggplot(df, aes(x=rps27l.stoich, y=ygene.umi.norm))
  # plot(p0 
  #      + geom_point() 
  #      + geom_smooth(method="lm", se=FALSE, fullrange=TRUE, color="gray50")
  #      + theme_classic()
  #      + theme(legend.position='none'))
  dev.off()

  }
```

